crates/accelerate/src/circuit_library/pauli_evolution.rs

```diff
@@ -0,0 +1,303 @@
+/// followed by a CX-chain and then a single Pauli-Z rotation on the last qubit. Then the CX-chain
+/// is uncomputed and the inverse basis transformation applied. E.g. for the evolution under the
+/// Pauli string XIYZ we have the circuit
+///                 ┌───┐┌───────┐┌───┐
+/// 0: ─────────────┤ X ├┤ Rz(2) ├┤ X ├───────────
+///    ┌──────┐┌───┐└─┬─┘└───────┘└─┬─┘┌───┐┌────┐
+/// 1: ┤ √Xdg ├┤ X ├──■─────────────■──┤ X ├┤ √X ├
+///    └──────┘└─┬─┘                   └─┬─┘└────┘
+/// 2: ──────────┼───────────────────────┼────────
+///     ┌───┐    │                       │  ┌───┐
+/// 3: ─┤ H ├────■───────────────────────■──┤ H ├─
+///     └───┘                               └───┘
+///
+/// Args:
+///     num_qubits: The number of qubits in the Hamiltonian.
+///     sparse_paulis: The Paulis to implement. Given in a sparse-list format with elements
+///         ``(pauli_string, qubit_indices, coefficient)``. An element of the form
+///         ``("IXYZ", [0,1,2,3], 0.2)``, for example, is interpreted in terms of qubit indices as
+///          I_q0 X_q1 Y_q2 Z_q3 and will use a RZ rotation angle of 0.4.
+///     insert_barriers: If ``true``, insert a barrier in between the evolution of individual
+///         Pauli terms.
+///     cx_fountain: If ``true``, implement the CX propagation as "fountain" shape, where each
+///         CX uses the top qubit as target. If ``false``, uses a "chain" shape, where CX in between
+///         neighboring qubits are used.
+///
+/// Returns:
+///     Circuit data for to implement the evolution.
+#[pyfunction]
+#[pyo3(signature = (num_qubits, sparse_paulis, insert_barriers=false, do_fountain=false))]
+pub fn py_pauli_evolution(
+    py: Python,
+    num_qubits: i64,
+    sparse_paulis: &Bound<PyList>,
+    insert_barriers: bool,
+    do_fountain: bool,
+) -> PyResult<CircuitData> {
+    let num_paulis = sparse_paulis.len();
+    let mut paulis: Vec<String> = Vec::with_capacity(num_paulis);
+    let mut indices: Vec<Vec<u32>> = Vec::with_capacity(num_paulis);
+    let mut times: Vec<Param> = Vec::with_capacity(num_paulis);
+    let mut global_phase = Param::Float(0.0);
+
+    for el in sparse_paulis.iter() {
+        let tuple = el.downcast::<PyTuple>()?;
+        let pauli = tuple.get_item(0)?.downcast::<PyString>()?.to_string();
+        let time = Param::extract_no_coerce(&tuple.get_item(2)?)?;
+
+        if pauli.as_str().chars().all(|p| p == 'i') {
+            global_phase = radd_param(global_phase, time, py); // apply factor -1 at the end
+            continue;
+        }
+
+        paulis.push(pauli);
+        times.push(time); // note we do not multiply by 2 here, this is done Python side!
+        indices.push(
+            tuple
+                .get_item(1)?
+                .downcast::<PyList>()?
+                .iter()
+                .map(|index| index.extract::<u32>())
+                .collect::<PyResult<_>>()?,
```

Comment:
Should we directly extract to a `Vec<u32>` here?
```suggestion
            tuple
                .get_item(1)?
                .extract::<Vec<u32>>()?,
```

Reply:
Oh yes -- I just recently learned that one can directly extract `Vec`s -- thanks!